simple/bug_world/bug.py

```diff
@@ -125,23 +120,49 @@ def grow(self):
     self.size += growth
     self.age += 1
     self.food_supply = 0
-    too_thin = self.size < self._calc_size_thres()
     return {
         "consumed_food": consumed,
-        "is_live": self.age < self._AGE_LIMIT and not too_thin,
+        "is_live": self.age < self._AGE_LIMIT and not self.is_too_thin(),
         "bug": self,
     }
 
   def choose_direction(self):
     """
         if the current spot is not the food spot, bug will move towards
         the food.
-        Bug knows the food potential in all four directions.
+        Bug knows the food potential in all eight directions.
         """
-    max = -1
-    dir = [0, 1]
+    strongest_smell = -1
+    strongest_smell_dir = [0, 1]
     for idx, d in enumerate(self.DIRS):
-      if self._vision[idx] > max:
-        max = self._vision[idx]
-        dir = d
-    return dir
+      if self._vision[idx] > strongest_smell:
+        strongest_smell = self._vision[idx]
+        strongest_smell_dir = d
+    return strongest_smell_dir
+
+  def will_reproduce(self):
+    """
+      if the bug's size is large than a certain threshold it will split
+        into two bugs.
+      we can add more conditions for reproduction later
+    """
+    REPRODUCE_SIZE_THRESHOLD = 20
+    return self.size > REPRODUCE_SIZE_THRESHOLD
+
+  def reproduce(self):
+    """
+        bug has a simple reproduction rule - splitting
+        the two new child bugs will start with age=0 and size equal to half
```

Comment:
I'm afraid this will make bugs after a few generations able to grow infinitely large. Maybe keep the mother at own age and size = current -1. The new on with certain initial size and age=0.